README.md
# libzbase32

libzbase32 is a `no_std` compatible crate that supports encoding and
decoding data in the z-base-32 format, as specified
[here](https://philzimmermann.com/docs/human-oriented-base-32-encoding.txt).

Z-base-32 is intended to be easier for a human to work with than regular
Base32 specified by [RFC 4658](https://datatracker.ietf.org/doc/html/rfc4648).

Some of the key differences:

* Z-base-32 a different alphabet ("ybndrfg8ejkmcpqxot1uwisza345h769") which
  consists of all lower-case letters (this library will accept lower-case
  or uppercase letters when decoding). The alphabet was chosen to make
  easier to use character appear more frequently in the output.

* Z-base-32 that the parties encoding and decoding z-base-32 values have
  some mechanism to agree on the length of the data. z-base-32 never
  includes padding characters (eg: "=") in order to keep the representation
  more compact.

* With Z-base-32, data lengths are specified in bits. This allows for more compact
  encodings. For example, in z-base-32, a 5 bit value can be encoded
  into a single character; while base32 would produce an
  8 character encoded value (of which 6 characters are padding bytes).

## Documentation

Modules documentation is available [here](https://docs.rs/libzbase32).

## License

This project is licensed under either of

* Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or
  <https://www.apache.org/licenses/LICENSE-2.0>)
* MIT license ([LICENSE-MIT](LICENSE-MIT) or
  <https://opensource.org/licenses/MIT>)

at your option.


CHANGES.md
2.0.1
=====

* Fix imports to make no_std mode actually work.

2.0.0
=====

* Rework the Error interfaces to make it easier to differentiate
  between errors related to bad input and errors related to incorrect
  usage.

1.1.0
=====

* Add is_last_quintet_valid() and is_last_octet_valid() functions which can
  be used to figure out if the final input symbol is valid given the number
  of bits. This can be useful for finding and correcting errors.


src/lib.rs
//! # libzbase32
//!
//! libzbase32 is a `no_std` compatible crate that supports encoding and
//! decoding data in the z-base-32 format, as specified
//! [here](https://philzimmermann.com/docs/human-oriented-base-32-encoding.txt).
//!
//! Z-base-32 is intended to be easier for a human to work with than regular
//! Base32 specified by [RFC 4658](https://datatracker.ietf.org/doc/html/rfc4648).
//!
//! Some of the key differences:
//!
//! * Z-base-32 a different alphabet ("ybndrfg8ejkmcpqxot1uwisza345h769") which
//!   consists of all lower-case letters (this library will accept lower-case
//!   or uppercase letters when decoding). The alphabet was chosen to make
//!   easier to use character appear more frequently in the output.
//!
//! * Z-base-32 that the parties encoding and decoding z-base-32 values have
//!   some mechanism to agree on the length of the data. z-base-32 never
//!   includes padding characters (eg: "=") in order to keep the representation
//!   more compact.
//!
//! * With Z-base-32, data lengths are specified in bits. This allows for more compact
//!   encodings. For example, in z-base-32, a 5 bit value can be encoded
//!   into a single character; while base32 would produce an
//!   8 character encoded value (of which 6 characters are padding bytes).
//!
//! ## Usage
//!
//! There are two APIs that can be used - the high-level API and the low-level API.
//! The high-level API is a little more convenient to use and should
//! generally be used when possible - but is unavailable in no_std mode. The
//! low-level API allows for no-allocation operation as well as two-step operations
//! in which transformations between octets <-> quintets and quintets <-> characters
//! are separate operations - which can be useful for specialized use cases.
//!
//! When using either API, its important to remember that z-base-32 is big-endian
//! oriented. As such, if you encode a single bit, z-base-32 will encode
//! the _highest bit_ of the input byte.
//!
//! When encoding or decoding, if the input value includes non-zero bits past
//! the number of bits specified in the operation, an Error be returned. For example,
//! encoding a single bit for the input value 0x01 will fail.
//!
//! Most fallible operations return an Error value of the same type, [`ZBase32Error`].
//! This is a mostly opaque type that only allows you to differentiate between an
//! error in the input value or an error in using the interfaces. More information
//! about the cause of the error can be retrieved by using the
//! [`Debug::fmt`](std::fmt::Debug::fmt) or [`Display::fmt`](std::fmt::Display::fmt)
//! functions.
//!
//! ## High-level API
//!
//! The high-level API consists of the functions [`encode`] and
//! its reverse, [`decode`].
//!
//! Example:
//!
//! ```
//! use libzbase32::{ZBase32Error, encode, decode};
//!
//! # fn main() {
//! const DATA: &'static [u8] = &[0, 44, 55, 128];
//!
//! let mut encoded = String::new();
//! encode(DATA, &mut encoded, 25).expect("Encoding failed!");
//!
//! assert_eq!(&encoded, "yysdx");
//!
//! let mut decoded = Vec::new();
//! decode(&encoded, &mut decoded, 25).expect("Decoding failed!");
//!
//! assert_eq!(&decoded, DATA);
//! # }
//! ```
//!
//! ## Low-level API
//!
//! The low-level API is found in the [`low_level_encode`] and [`low_level_decode`] modules.
//!
//! Unlike the high-level API:
//!
//! * The low-level API does not allocate.
//!
//! * Unlike The high-level API, all input and output types are `&[u8]` and its up
//!   to the caller to differentiate between arrays of bytes (octets), quintets, or
//!   ASCII character values.
//!
//! * The low-level API supports two-step operations. Eg, when encoding, the low-level API allows
//!   users to first transform the input from octets in quintet values (integers
//!   in the range 0-31) and then later transform from quintet values into
//!   encoded characters. This can be handy for specialized applications.
//!
//! * Single-step operations are also supported via the
//!   [`encode_slices`](low_level_encode::encode_slices) and
//!   [`decode_slices`](low_level_decode::decode_slices) which function similarly
//!   to the functions in the high-level API - but require the caller to setup
//!   an appropriate output buffer.
//!
//! ## No_std
//!
//! No_std mode may be activated by disabling the "std" feature. In this
//! mode, only the low-level interfaces are available.
//!
//! ## License
//
//! This project is licensed under either of
//!
//! * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or
//!   <https://www.apache.org/licenses/LICENSE-2.0>)
//! * MIT license ([LICENSE-MIT](LICENSE-MIT) or
//!   <https://opensource.org/licenses/MIT>)
//!
//! at your option.

#![cfg_attr(not(feature = "std"), no_std)]

mod decode_impl;
mod encode_impl;
mod error;
mod stateful_decoder;
mod stateful_encoder;
mod tables;
#[cfg(test)]
mod test_data;
mod util;

pub use error::{InputErrorCause, UsageError, UsageErrorCause, ZBase32Error};

#[cfg(feature = "std")]
pub use decode_impl::decode;

#[cfg(feature = "std")]
pub use encode_impl::encode;

pub mod low_level_decode {
    //! Low-level decoding functionality
    pub use crate::decode_impl::{
        character_to_quintet, decode_slices, is_last_quintet_valid, quintets_to_octets,
    };
    pub use crate::util::required_octets_buffer_len;
}

pub mod low_level_encode {
    //! Low-level encoding functionality
    pub use crate::encode_impl::{
        encode_slices, is_last_octet_valid, octets_to_quintets, quintet_to_character,
    };
    pub use crate::util::required_quintets_buffer_len;
}


src/util.rs
use crate::error::bits_overflow;
use crate::UsageError;

const fn u64_to_usize(val: u64) -> Option<usize> {
    if usize::BITS >= u64::BITS {
        Some(val as usize)
    } else {
        // In this case, u64::BITS > usize::BITS - so, any usize
        // value can be safely casted to a u64.
        let max_usize: u64 = usize::MAX as u64;
        if val <= max_usize {
            Some(val as usize)
        } else {
            None
        }
    }
}

/// Calculate the number of octets that are required to hold the specified
/// number of bits.
///
/// This function will return an Err value if the specified number of bits
/// would result in needing more than [`usize::MAX`] octets.
pub const fn required_octets_buffer_len(bits: u64) -> Result<usize, UsageError> {
    let remainder = bits % 8;
    let needed_octets = bits / 8 + if remainder == 0 { 0 } else { 1 };
    if let Some(result) = u64_to_usize(needed_octets) {
        Ok(result)
    } else {
        Err(bits_overflow())
    }
}

/// Calculate the number of quintets tht are required to hold the specified
/// number of bits.
///
/// This function will return an Err value if the specified number of bits
/// would result in needing more than [`usize::MAX`] quintets.
pub const fn required_quintets_buffer_len(bits: u64) -> Result<usize, UsageError> {
    let remainder = bits % 5;
    let needed_octets = bits / 5 + if remainder == 0 { 0 } else { 1 };
    if let Some(result) = u64_to_usize(needed_octets) {
        Ok(result)
    } else {
        Err(bits_overflow())
    }
}


src/error.rs
use core::fmt::{Debug, Display, Formatter};

enum InputErrorType {
    InvalidCharacter,
    InvalidQuintet,
    TrailingNonZeroBits,
}

pub struct InputErrorCause {
    typ: InputErrorType,
}

impl Debug for InputErrorCause {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self.typ {
            InputErrorType::InvalidCharacter => write!(f, "Invalid character found in input."),
            InputErrorType::InvalidQuintet => write!(f, "Invalid quintet value found in input."),
            InputErrorType::TrailingNonZeroBits => {
                write!(f, "Trailing non-zero bits found in input.")
            }
        }
    }
}

impl Display for InputErrorCause {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        Debug::fmt(self, f)
    }
}

enum UsageErrorType {
    InputBufferDoesntMatchBits,
    OutputBufferDoesntMatchBits,
    BitsOverflow,
}

pub struct UsageErrorCause {
    typ: UsageErrorType,
}

impl Debug for UsageErrorCause {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self.typ {
            UsageErrorType::InputBufferDoesntMatchBits => {
                write!(
                    f,
                    "The input buffer size doesn't agree with the provided bits value"
                )
            }
            UsageErrorType::OutputBufferDoesntMatchBits => {
                write!(
                    f,
                    "The output buffer size doesn't agree with the provided bits value"
                )
            }
            UsageErrorType::BitsOverflow => {
                write!(f, "The value for bits was too large for the platform usize")
            }
        }
    }
}

impl Display for UsageErrorCause {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        Debug::fmt(self, f)
    }
}

/// Common error type used by all fallible operations where bad input could be the cause
///
/// By design, this type is mostly opaque - with the exception
/// that its possible to differentiate between errors with input
/// data and other types of errors. The [`Debug`] or [`Display`]
/// implementations can be used to format a more specific error
/// message.
pub enum ZBase32Error {
    /// An InputError indicates that an input array contained an invalid
    /// value. For example, a non-zbase32 character being passed to one of
    /// the decode methods.
    InputError(InputErrorCause),

    /// A UsageError indicates an error outside of an invalid input value.
    UsageError(UsageErrorCause),
}

impl Debug for ZBase32Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            ZBase32Error::InputError(cause) => write!(f, "Input Error: {}", cause),
            ZBase32Error::UsageError(cause) => write!(f, "Usage Error: {}", cause),
        }
    }
}

impl Display for ZBase32Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        Debug::fmt(self, f)
    }
}

#[cfg(feature = "std")]
impl std::error::Error for ZBase32Error {}

impl From<UsageError> for ZBase32Error {
    fn from(err: UsageError) -> Self {
        ZBase32Error::UsageError(err.0)
    }
}

/// A UsageError indicates an error outside of an invalid input value.
///
/// By design, this type is mostly opaque. The [`Debug`] or [`Display`]
/// implementations can be used to format a more specific error
/// message.
pub struct UsageError(pub UsageErrorCause);

impl Debug for UsageError {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        write!(f, "Usage Error: {}", self.0)
    }
}

impl Display for UsageError {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        Debug::fmt(self, f)
    }
}

#[cfg(feature = "std")]
impl std::error::Error for UsageError {}

pub const fn invalid_character() -> ZBase32Error {
    ZBase32Error::InputError(InputErrorCause {
        typ: InputErrorType::InvalidCharacter,
    })
}

pub const fn invalid_quintet() -> ZBase32Error {
    ZBase32Error::InputError(InputErrorCause {
        typ: InputErrorType::InvalidQuintet,
    })
}

pub const fn trailing_nonzero_bits() -> ZBase32Error {
    ZBase32Error::InputError(InputErrorCause {
        typ: InputErrorType::TrailingNonZeroBits,
    })
}

pub const fn input_buffer_doesnt_match_bits() -> UsageError {
    UsageError(UsageErrorCause {
        typ: UsageErrorType::InputBufferDoesntMatchBits,
    })
}

pub const fn output_buffer_doesnt_match_bits() -> UsageError {
    UsageError(UsageErrorCause {
        typ: UsageErrorType::OutputBufferDoesntMatchBits,
    })
}

pub const fn bits_overflow() -> UsageError {
    UsageError(UsageErrorCause {
        typ: UsageErrorType::BitsOverflow,
    })
}


src/decode_impl.rs
use crate::error::{
    input_buffer_doesnt_match_bits, invalid_character, output_buffer_doesnt_match_bits,
};
use crate::stateful_decoder::{
    quintet_has_valid_trailing_bits, HaveOctets, NeedQuintets, NextOctetResult,
    ProvideQuintetResult,
};
use crate::tables::{CHARACTER_MIN_VALUE, CHARACTER_TO_QUINTET};
use crate::util::{required_octets_buffer_len, required_quintets_buffer_len};
use crate::ZBase32Error;
use core::iter::Peekable;

enum QuintetsToOctetsIterState {
    Initial(NeedQuintets),
    HaveOctets(HaveOctets),
}

struct QuintetsToOctetsIter<I>
where
    I: Iterator<Item = Result<u8, ZBase32Error>>,
{
    quintet_iter: Peekable<I>,
    state: Option<QuintetsToOctetsIterState>,
}

impl<I> QuintetsToOctetsIter<I>
where
    I: Iterator<Item = Result<u8, ZBase32Error>>,
{
    fn new(quintet_iter: I, need_quintets: NeedQuintets) -> QuintetsToOctetsIter<I> {
        QuintetsToOctetsIter {
            quintet_iter: quintet_iter.peekable(),
            state: Some(QuintetsToOctetsIterState::Initial(need_quintets)),
        }
    }
}

fn refill<I>(
    quintet_iter: &mut Peekable<I>,
    mut need_quintets: NeedQuintets,
) -> Result<Option<QuintetsToOctetsIterState>, ZBase32Error>
where
    I: Iterator<Item = Result<u8, ZBase32Error>>,
{
    loop {
        let quintet = quintet_iter.next().unwrap()?;
        let last_quintet = quintet_iter.peek().is_none();
        match need_quintets.provide_quintet(quintet, last_quintet)? {
            ProvideQuintetResult::NeedQuintets(need_more) => need_quintets = need_more,
            ProvideQuintetResult::HaveOctets(have_octets) => {
                return Ok(Some(QuintetsToOctetsIterState::HaveOctets(have_octets)))
            }
        }
    }
}

impl<I> Iterator for QuintetsToOctetsIter<I>
where
    I: Iterator<Item = Result<u8, ZBase32Error>>,
{
    type Item = Result<u8, ZBase32Error>;

    fn next(&mut self) -> Option<<Self as Iterator>::Item> {
        loop {
            if self.state.is_none() {
                return None;
            }

            match self.state.take().unwrap() {
                // The first time we are called, we start in the initial state
                // and have to read some quintets from the underlying iterator.
                // After doing this, we'll loop and never end up in this state
                // again.
                QuintetsToOctetsIterState::Initial(need_quintets) => {
                    if self.quintet_iter.peek().is_some() {
                        match refill(&mut self.quintet_iter, need_quintets) {
                            Ok(new_state) => self.state = new_state,
                            Err(err) => return Some(Err(err)),
                        }
                    } else {
                        // Do nothing - this means that the input iterator was empty -
                        // which shouldn't actually be possible.
                    }
                }

                // If we are in the HaveOctets state, it means we can try to return
                // already processed octet values. If we run out of octets, we will
                // need to use the refill method to process more quintets into
                // octets. If we hit the end of the input, we'll leave self.state
                // as None as a signal to complete.
                QuintetsToOctetsIterState::HaveOctets(have_octets) => {
                    match have_octets.next_octet() {
                        NextOctetResult::Octet(octet, have_octets) => {
                            self.state = Some(QuintetsToOctetsIterState::HaveOctets(have_octets));
                            return Some(Ok(octet));
                        }
                        NextOctetResult::NeedQuintets(need_quintets) => {
                            match refill(&mut self.quintet_iter, need_quintets) {
                                Ok(new_state) => self.state = new_state,
                                Err(err) => return Some(Err(err)),
                            }
                        }
                        NextOctetResult::Complete => {}
                    }
                }
            }
        }
    }
}

/// Convert a character code value (such as "y") to its integer
/// value (such as 0).
pub fn character_to_quintet(character: u8) -> Result<u8, ZBase32Error> {
    if character < CHARACTER_MIN_VALUE {
        return Err(invalid_character());
    } else if (character - CHARACTER_MIN_VALUE) as usize >= CHARACTER_TO_QUINTET.len() {
        return Err(invalid_character());
    }
    let val = CHARACTER_TO_QUINTET[(character - CHARACTER_MIN_VALUE) as usize];
    if val == 255 {
        return Err(invalid_character());
    }
    Ok(val)
}

fn calc_last_quintet_bits(bits: u64) -> Option<u8> {
    if bits == 0 {
        None
    } else {
        match bits % 5 {
            0 => Some(5),
            x => Some(x as u8),
        }
    }
}

/// Determine if the last quintet is, given the number of bits to decode
pub fn is_last_quintet_valid(bits: u64, quintet: u8) -> bool {
    if let Some(last_quintet_bits) = calc_last_quintet_bits(bits) {
        quintet <= 31 && quintet_has_valid_trailing_bits(last_quintet_bits, quintet)
    } else {
        false
    }
}

/// Convert a buffer of quintet integer values (ie, integers of the
/// range 0-31) into a buffer of octet (byte) values.
///
/// The length of `in_quintets` must match the value returned by
/// [`required_quintets_buffer_len`] and the length of `out_octets`
/// must match the value returned by [`required_octets_buffer_len`].
pub fn quintets_to_octets(
    in_quintets: &[u8],
    out_octets: &mut [u8],
    bits: u64,
) -> Result<(), ZBase32Error> {
    if in_quintets.len() != required_quintets_buffer_len(bits)? {
        return Err(input_buffer_doesnt_match_bits().into());
    }
    if out_octets.len() != required_octets_buffer_len(bits)? {
        return Err(output_buffer_doesnt_match_bits().into());
    }
    let last_quintet_bits = if let Some(x) = calc_last_quintet_bits(bits) {
        x
    } else {
        return Ok(());
    };

    let octet_iter = QuintetsToOctetsIter::new(
        in_quintets.iter().map(|&x| Ok(x)),
        NeedQuintets::new(last_quintet_bits),
    );

    for (out, next_octet) in out_octets.iter_mut().zip(octet_iter) {
        *out = next_octet?;
    }

    Ok(())
}

/// Decode a buffer of characters to a buffer of octets (bytes).
///
/// This method is the low-level equivalent of the [`decode`] method.
/// Unlike that method, this method does not allocate and is usable
/// in `no_std` mode.
///
/// This method is equivalent to first using [`character_to_quintet`] to
/// convert all character values to quintet integers and then using
/// [`quintets_to_octets`] to convert to octet (byte) values. This method,
/// however, may be more efficient as it is able to do both operations
/// together in a single pass over the input buffer.
///
/// The length of `in_characters` must match the value returned by
/// [`required_quintets_buffer_len`] and the length of `out_octets`
/// must match the value returned by [`required_octets_buffer_len`].
pub fn decode_slices(
    in_characters: &[u8],
    out_octets: &mut [u8],
    bits: u64,
) -> Result<(), ZBase32Error> {
    if in_characters.len() != required_quintets_buffer_len(bits)? {
        return Err(input_buffer_doesnt_match_bits().into());
    }
    if out_octets.len() != required_octets_buffer_len(bits)? {
        return Err(output_buffer_doesnt_match_bits().into());
    }
    let last_quintet_bits = if let Some(x) = calc_last_quintet_bits(bits) {
        x
    } else {
        return Ok(());
    };

    let octet_iter = QuintetsToOctetsIter::new(
        in_characters.iter().map(|&x| character_to_quintet(x)),
        NeedQuintets::new(last_quintet_bits),
    );

    for (out, next_octet) in out_octets.iter_mut().zip(octet_iter) {
        *out = next_octet?;
    }

    Ok(())
}

/// Decode a slice of characters to a [`Vec`] of octets (bytes).
///
/// The output octets will be appended to `output`.
///
/// The length of `input` must match the value returned by
/// [`required_quintets_buffer_len`]. The `output` buffer will be
/// extended by to accommodate the output.
///
/// This method is not available in `no_std` mode.
#[cfg(feature = "std")]
pub fn decode(input: &str, output: &mut Vec<u8>, bits: u64) -> Result<(), ZBase32Error> {
    if input.len() != required_quintets_buffer_len(bits)? {
        return Err(input_buffer_doesnt_match_bits().into());
    }
    let last_quintet_bits = if let Some(x) = calc_last_quintet_bits(bits) {
        x
    } else {
        return Ok(());
    };

    let needed_octets = required_octets_buffer_len(bits)?;
    let start = output.len();
    output.extend(std::iter::repeat(0).take(needed_octets));
    let output_buff = &mut output[start..];

    let octet_iter = QuintetsToOctetsIter::new(
        input.as_bytes().iter().map(|&x| character_to_quintet(x)),
        NeedQuintets::new(last_quintet_bits),
    );

    for (out, next_octet) in output_buff.iter_mut().zip(octet_iter) {
        *out = next_octet?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::decode;
    use crate::test_data::{TestCase, RANDOM_TEST_DATA, STANDARD_TEST_DATA};

    fn run_tests(test_cases: &[TestCase]) {
        let mut buffer = Vec::new();
        for test in test_cases {
            buffer.clear();
            decode(test.encoded, &mut buffer, test.bits).unwrap();
            assert_eq!(&buffer[..], test.unencoded);
        }
    }

    #[test]
    fn test_decode_standard() {
        run_tests(STANDARD_TEST_DATA);
    }

    #[test]
    fn test_decode_random() {
        run_tests(RANDOM_TEST_DATA);
    }
}


src/encode_impl.rs
use crate::error::{
    input_buffer_doesnt_match_bits, invalid_quintet, output_buffer_doesnt_match_bits,
};
use crate::stateful_encoder::{
    octet_has_valid_trailing_bits, HaveQuintets, NeedOctets, NextQuintetResult, ProvideOctetResult,
};
use crate::tables::QUINTET_TO_CHARACTER;
use crate::util::{required_octets_buffer_len, required_quintets_buffer_len};
use crate::ZBase32Error;
use core::iter::Peekable;

enum OctetsToQuintetsIterState {
    Initial(NeedOctets),
    HaveQuintets(HaveQuintets),
}

struct OctetsToQuintetsIter<I>
where
    I: Iterator<Item = u8>,
{
    octet_iter: Peekable<I>,
    state: Option<OctetsToQuintetsIterState>,
}

impl<I> OctetsToQuintetsIter<I>
where
    I: Iterator<Item = u8>,
{
    fn new(octet_iter: I, need_octets: NeedOctets) -> OctetsToQuintetsIter<I> {
        OctetsToQuintetsIter {
            octet_iter: octet_iter.peekable(),
            state: Some(OctetsToQuintetsIterState::Initial(need_octets)),
        }
    }
}

fn refill<I>(
    octet_iter: &mut Peekable<I>,
    mut need_octets: NeedOctets,
) -> Result<Option<OctetsToQuintetsIterState>, ZBase32Error>
where
    I: Iterator<Item = u8>,
{
    loop {
        let octet = octet_iter.next().unwrap();
        let last_octet = octet_iter.peek().is_none();
        match need_octets.provide_octet(octet, last_octet)? {
            ProvideOctetResult::NeedOctets(need_more) => need_octets = need_more,
            ProvideOctetResult::HaveQuintets(have_quintets) => {
                return Ok(Some(OctetsToQuintetsIterState::HaveQuintets(have_quintets)))
            }
        }
    }
}

impl<I> Iterator for OctetsToQuintetsIter<I>
where
    I: Iterator<Item = u8>,
{
    type Item = Result<u8, ZBase32Error>;

    fn next(&mut self) -> Option<<Self as Iterator>::Item> {
        loop {
            if self.state.is_none() {
                return None;
            }

            match self.state.take().unwrap() {
                // The first time we are called, we start in the initial state
                // and have to read some quintets from the underlying iterator.
                // After doing this, we'll loop and never end up in this state
                // again.
                OctetsToQuintetsIterState::Initial(need_octets) => {
                    if self.octet_iter.peek().is_some() {
                        match refill(&mut self.octet_iter, need_octets) {
                            Ok(new_state) => self.state = new_state,
                            Err(err) => return Some(Err(err)),
                        }
                    } else {
                        // Do nothing - this means that the input iterator was empty -
                        // which shouldn't actually be possible.
                    }
                }

                // If we are in the HaveOctets state, it means we can try to return
                // already processed octet values. If we run out of octets, we will
                // need to use the refill method to process more quintets into
                // octets. If we hit the end of the input, we'll leave self.state
                // as None as a signal to complete.
                OctetsToQuintetsIterState::HaveQuintets(have_quintets) => {
                    match have_quintets.next_quintet() {
                        NextQuintetResult::Quintet(quintet, have_quintets) => {
                            self.state =
                                Some(OctetsToQuintetsIterState::HaveQuintets(have_quintets));
                            return Some(Ok(quintet));
                        }
                        NextQuintetResult::NeedOctets(need_octets) => {
                            match refill(&mut self.octet_iter, need_octets) {
                                Ok(new_state) => self.state = new_state,
                                Err(err) => return Some(Err(err)),
                            }
                        }
                        NextQuintetResult::Complete => {}
                    }
                }
            }
        }
    }
}

/// Convert a quintet integer value (such as "0") to its character
/// value (such as "y").
pub fn quintet_to_character(quintet: u8) -> Result<u8, ZBase32Error> {
    if quintet as usize > QUINTET_TO_CHARACTER.len() {
        return Err(invalid_quintet());
    }
    Ok(QUINTET_TO_CHARACTER[quintet as usize])
}

fn calc_last_octet_bits(bits: u64) -> Option<u8> {
    if bits == 0 {
        None
    } else {
        match bits % 8 {
            0 => Some(8),
            x => Some(x as u8),
        }
    }
}

/// Determine if the last octet is valid, given the number of bits to encode
pub fn is_last_octet_valid(bits: u64, octet: u8) -> bool {
    if let Some(last_octet_bits) = calc_last_octet_bits(bits) {
        octet_has_valid_trailing_bits(last_octet_bits, octet)
    } else {
        false
    }
}

/// Convert a buffer of octets (bytes) into a buffer of quintet values (ie, integers between 0-31).
///
/// The length of `in_octets` must match the value returned by
/// [`required_octets_buffer_len`] and the length of `out_quintets`
/// must match the value returned by [`required_quintets_buffer_len`].
pub fn octets_to_quintets(
    in_octets: &[u8],
    out_quintets: &mut [u8],
    bits: u64,
) -> Result<(), ZBase32Error> {
    if in_octets.len() != required_octets_buffer_len(bits)? {
        return Err(input_buffer_doesnt_match_bits().into());
    }
    if out_quintets.len() != required_quintets_buffer_len(bits)? {
        return Err(output_buffer_doesnt_match_bits().into());
    }
    let last_octet_bits = if let Some(x) = calc_last_octet_bits(bits) {
        x
    } else {
        return Ok(());
    };

    let quintet_iter = OctetsToQuintetsIter::new(
        in_octets.iter().map(|&x| x),
        NeedOctets::new(last_octet_bits),
    );

    for (out_quintet, in_quintet) in out_quintets.iter_mut().zip(quintet_iter) {
        *out_quintet = in_quintet?;
    }

    Ok(())
}

/// Encode a buffer of octets (bytes) to a buffer of characters.
///
/// This method is the low-level equivalent of the [`encode`] method.
/// Unlike that method, this method does not allocate and is usable
/// in `no_std` mode.
///
/// This method is equivalent to first using [`octets_to_quintets`] to
/// convert all octet (byte) values to quintet integers and then using
/// [`quintet_to_character`] to convert those values to character values. This method,
/// however, may be more efficient as it is able to do both operations
/// together in a single pass over the input buffer.
///
/// The length of `in_octets` must match the value returned by
/// [`required_octets_buffer_len`] and the length of `out_characters`
/// must match the value returned by [`required_quintets_buffer_len`].
pub fn encode_slices(
    in_octets: &[u8],
    out_characters: &mut [u8],
    bits: u64,
) -> Result<(), ZBase32Error> {
    if in_octets.len() != required_octets_buffer_len(bits)? {
        return Err(input_buffer_doesnt_match_bits().into());
    }
    if out_characters.len() != required_quintets_buffer_len(bits)? {
        return Err(output_buffer_doesnt_match_bits().into());
    }
    let last_octet_bits = if let Some(x) = calc_last_octet_bits(bits) {
        x
    } else {
        return Ok(());
    };

    let quintet_iter = OctetsToQuintetsIter::new(
        in_octets.iter().map(|&x| x),
        NeedOctets::new(last_octet_bits),
    );

    for (out_quintet, in_quintet) in out_characters.iter_mut().zip(quintet_iter) {
        *out_quintet = quintet_to_character(in_quintet?)?;
    }

    Ok(())
}

/// Encode a slice of octets (bytes) to a [`String`].
///
/// The output characters will be appended to `output`.
///
/// The length of `input` must match the value returned by
/// [`required_octets_buffer_len`]. The `output` buffer will be
/// extended by to accommodate the output.
///
/// This method is not available in `no_std` mode.
#[cfg(feature = "std")]
pub fn encode(input: &[u8], output: &mut String, bits: u64) -> Result<(), ZBase32Error> {
    if input.len() != required_octets_buffer_len(bits)? {
        return Err(input_buffer_doesnt_match_bits().into());
    }
    let last_octet_bits = if let Some(x) = calc_last_octet_bits(bits) {
        x
    } else {
        return Ok(());
    };

    let needed_quintets = required_quintets_buffer_len(bits)?;
    output.reserve(needed_quintets);

    let quintet_iter =
        OctetsToQuintetsIter::new(input.iter().map(|&x| x), NeedOctets::new(last_octet_bits));

    for quintet in quintet_iter {
        output.push(quintet_to_character(quintet?)? as char);
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::encode;
    use crate::test_data::{TestCase, RANDOM_TEST_DATA, STANDARD_TEST_DATA};

    fn run_tests(test_cases: &[TestCase]) {
        let mut buffer = String::new();
        for test in test_cases {
            buffer.clear();
            encode(test.unencoded, &mut buffer, test.bits).unwrap();
            assert_eq!(&buffer, test.encoded);
        }
    }

    #[test]
    fn test_encode_standard() {
        run_tests(STANDARD_TEST_DATA);
    }

    #[test]
    fn test_encode_random() {
        run_tests(RANDOM_TEST_DATA);
    }
}


src/tables.rs
// Maps from character code to their value in the z-base-32 scheme
// Table starts from character code 49 ('1')
pub const CHARACTER_MIN_VALUE: u8 = 49;
pub const CHARACTER_TO_QUINTET: &'static [u8] = &[
    18, // '1'
    255, 25, // '3'
    26, // '4'
    27, // '5'
    30, // '6'
    29, // '7'
    7,  // '8'
    31, // '9'
    255, 255, 255, 255, 255, 255, 255, 24, // 'A'
    1,  // 'B'
    12, // 'C'
    3,  // 'D'
    8,  // 'E'
    5,  // 'F'
    6,  // 'G'
    28, // 'H'
    21, // 'I'
    9,  // 'J'
    10, // 'K'
    255, 11, // 'M'
    2,  // 'N'
    16, // 'O'
    13, // 'P'
    14, // 'Q'
    4,  // 'R'
    22, // 'S'
    17, // 'T'
    19, // 'U'
    255, 20, // 'W'
    15, // 'X'
    0,  // 'Y'
    23, // 'Z'
    255, 255, 255, 255, 255, 255, 24, // 'a'
    1,  // 'b'
    12, // 'c'
    3,  // 'd'
    8,  // 'e'
    5,  // 'f'
    6,  // 'g'
    28, // 'h'
    21, // 'i'
    9,  // 'j'
    10, // 'k'
    255, 11, // 'm'
    2,  // 'n'
    16, // 'o'
    13, // 'p'
    14, // 'q'
    4,  // 'r'
    22, // 's'
    17, // 't'
    19, // 'u'
    255, 20, // 'w'
    15, // 'x'
    0,  // 'y'
    23, // 'z'
];

pub const QUINTET_TO_CHARACTER: &'static [u8] = &[
    b'y', b'b', b'n', b'd', b'r', b'f', b'g', b'8', b'e', b'j', b'k', b'm', b'c', b'p', b'q', b'x',
    b'o', b't', b'1', b'u', b'w', b'i', b's', b'z', b'a', b'3', b'4', b'5', b'h', b'7', b'6', b'9',
];
